modules/api_reference/tests/src/Kernel/ParseTest.php

```diff
@@ -0,0 +1,219 @@
+use Symfony\Component\HttpFoundation\RequestStack;
+
+/**
+ * Tests the parsing and validating edge-case swagger/openapi files.
+ */
+final class ParseTest extends KernelTestBase {
+
+  /**
+   * Mock logger that does nothing.
+   *
+   * @var \Drupal\Core\Logger\LoggerChannelInterface
+   */
+  private $mockLogger;
+
+  /**
+   * Mock cache that does nothing.
+   *
+   * @var \Drupal\Core\Cache\CacheBackendInterface
+   */
+  private $mockCache;
+
+  /**
+   * Creates a Swagger plugin.
+   *
+   * @return \Drupal\devportal_api_reference\Plugin\Reference\Swagger
+   *   Instance.
+   */
+  private function createSwaggerPlugin(): Swagger {
+    return new Swagger([], '', [], $this->mockCache, $this->mockLogger);
+  }
+
+  /**
+   * Creates an OpenApi3 plugin.
+   *
+   * @return \Drupal\devportal_api_reference\Plugin\Reference\OpenApi3
+   *   Instance.
+   */
+  private function createOpenApi3Plugin(): OpenApi3 {
+    return new OpenApi3([], '', [], $this->mockCache, $this->mockLogger);
+  }
+
+  /**
+   * Lists the files in a subdirectory of fixtures/specials.
+   *
+   * @param string $subdir
+   *   Subdirectory name.
+   *
+   * @return array
+   *   Directory contents in a format that is accepted by PhpUnit's data
+   *   provider. The key is the file name, the value is an array with a single
+   *   element that is the full path of the file.
+   */
+  private function listFiles(string $subdir): array {
+    $dir = realpath(__DIR__ . "/../../fixtures/specials/$subdir");
+
+    $files = [];
+
+    if ($dh = @opendir($dir)) {
+      while (($file = readdir($dh)) !== FALSE) {
+        $path = "{$dir}/{$file}";
+        if (strpos($file, '.') !== 0 && filetype($path) === 'file') {
+          $files[$file] = [$path];
+        }
+      }
+      closedir($dh);
+    }
+
+    return $files;
+  }
+
+  /**
+   * Provides a list of special swagger files.
+   *
+   * @return array
+   *   List of files.
+   *
+   * @see \Drupal\Tests\devportal_api_reference\Kernel\ParseTest::listFiles()
+   */
+  public function swaggerFiles(): array {
+    return $this->listFiles('swagger');
+  }
+
+  /**
+   * Provides a list of special OpenApi3 files.
+   *
+   * @return array
+   *   List of files.
+   *
+   * @see \Drupal\Tests\devportal_api_reference\Kernel\ParseTest::listFiles()
+   */
+  public function openApi3Files(): array {
+    return $this->listFiles('openapi3');
+  }
+
+  /**
+   * Tests a Swagger file.
+   *
+   * @param string $filepath
+   *   Full path to the test file.
+   *
+   * @dataProvider swaggerFiles
+   */
+  public function testSwaggerFile(string $filepath): void {
+    self::assertNotNull($this->createSwaggerPlugin()->parse($filepath));
+  }
+
+  /**
+   * Tests an OpenApi3 file.
+   *
+   * @param string $filepath
+   *   Full path to the test file.
+   *
+   * @dataProvider openApi3Files
+   */
+  public function testOpenApi3File(string $filepath): void {
+    self::assertNotNull($this->createOpenApi3Plugin()->parse($filepath));
```

Comment:
Well, I am not sure how this test validates that the new `fixPattern()` method works properly, don't we need a Javascript test for that?

Reply:
No, because this check happens on the backend only.

Reply:
Okay, should not you validate the output returned by this function? Whether it escapes the regexp-s properly or not?

Reply:
What `parse()` can "return":

- swagger/openapi3 structure
- null if the validation fails
- exception if an error occoured

The `assertNotNull()` checks that a structure is returned. The validity of the structure is not important in this test, because that is being validated in the upload test.

Reply:
> null if the validation fails
> exception if an error occoured

Should not it throw an exception when the validation fails as well? (with some info about the reason of failure)

> The assertNotNull() checks that a structure is returned. The validity of the structure is not important in this test, because that is being validated in the upload test.
:ok_hand:

Reply:
`null` is returned when the preliminary check comes back as false. This preliminary check tries to determine if the file "looks like" it should be a validated with the given plugin. E.g. an obviously OpenAPI 3 file (which contains `openapi: 3.0.2`) won't get validated with the Swagger plugin.